### PR TITLE
feat: Dispatcher contracts & some tests

### PR DESCRIPTION
**Pull request by akhercha**
Resolves: #NA

## Work done
* Dispatcher contract
* Router contracts
* Registry tests
* Bump scarb

### CODE REVIEW

**Comment by JordyRo1** [comment]
lgtm

File: cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+
+            // [Effect] Add the asset class router to the storage
+            self.asset_class_routers.entry(asset_class).write(router);
+
+            // [Interaction] Storage updated event
+            self
+                .emit(
+                    AssetClassRouterAdded {
+                        sender: get_caller_address(),
+                        asset_class_id: asset_class_id,
```
Comment (EvolveArt) [inline cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo:147]: Why not emit the asset class ?

File: cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        ///
+        ///  3. Send the updates using the [dispatch] Hyperlane function.
+        fn dispatch(self: @ContractState, feed_ids: Span<FeedId>) -> HyperlaneMessageId {
+            // [Check] Assert that all feeds id are registered in the Registry
+            self.assert_all_feeds_exists(feed_ids.clone());
+
+            // [Effect] Add the number of feeds to update to the message
+            let mut update_message = BytesTrait::new_empty();
+            let nb_feeds_to_update: u32 = feed_ids.len();
+            update_message.append_u32(nb_feeds_to_update);
```
Comment (EvolveArt) [inline cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo]: u32 is probably too large we could reduce to u16

File: cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            // [Check] Contracts are not zero
+            assert(!summary_stats_address.is_zero(), errors::SUMMARY_STATS_IS_ZERO);
+
+            // [Check] Feed type id is valid
+            let feed_type = match FeedTypeTrait::from_id(feed_type_id) {
+                Result::Ok(f) => f,
+                Result::Err(e) => panic_with_felt252(e.into())
+            };
+            // [Check] Feed type variant & extract parameters
+            let (simple_data_type, aggregation_mode, duration) = match feed_type {
```
Comment (EvolveArt) [inline cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo:116]: This is a bit too complicated, I'd just have imagined that the parameters of the twap are given in the constructor and then used in the get_data function. No need to match on the variant or whatever

File: cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                        feed_type_id: feed_type_id,
+                        router_address: get_contract_address(),
+                    }
+                )
+        }
+    }
+
+    impl SummaryStatsWrapper of ISummaryStatsWrapper<ContractState> {
+        /// Calls calculate_volatility from the Summary Stats contract.
+        fn call_calculate_volatility(
```
Comment (EvolveArt) [inline cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo]: Why is that here ?

File: cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo (hunk 1 above)
Comment (akhercha) [reply cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo:147]: We can also but I think that if we return an Enum like that it will actually print the enum encoded to felt, right ? So basically the same as the id. Then I prefer it to be explicit and mark it as `asset_class_id`

File: cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo (hunk 2 above)
Comment (akhercha) [reply cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo]: True - what is the size in the Solidity parser? @JordyRo1 
I've found `u8` instead of `u16` but not sure

File: cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo (hunk 3 above)
Comment (akhercha) [reply cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo:116]: Mmmmh, I agree but I'm not sure either:

1. If we can indeed pass arbitrary parameters with the constructor, we would loose all traceability for which routers ***can*** exist,
2. By using only the `feed_type_id`, we're guaranteed to be able to use this id in the `AssetClassRouter` and find a unique contract.
3. It also mean that by looking at a documentation (or the code, as a developer), you can find all the possible `feed_type_id`  and it's a 1 <=> 1 matching with the routers that are (or can be) deployed, there can't be more definitions "hidden" or that we have "to guess" because the constructor allows anything.

But I agree that it makes the process of creating a new feed type more challenging (add the new variant in the types + the arm match in the router contract). I don't think we will create tons of them though.

any thoughts @JordyRo1 ?

File: cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo (hunk 4 above)
Comment (akhercha) [reply cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo]: The interface `ISummaryStatsWrapper` is the same for every feed types so we need to implement every functions. Any better idea?

EDIT: I can probably just remove this notion of "wrapper" and use a raw dispatcher. Wdyt?

File: cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo (hunk 2 above)
Comment (JordyRo1) [reply cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo]: yes, in solidity it's u8, to be safe I will upgrade to u16

File: cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo (hunk 4 above)
Comment (JordyRo1) [reply cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo]: Yes, I think we can just use the raw dispatcher here

File: cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo (hunk 3 above)
Comment (JordyRo1) [reply cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo:116]: I do agree with @akhercha on this point. Indeed, it makes the creation process more complicated, but a way better tracability, since we already know the `feed_type_id` used in the entire process. But since we will be the only allowed to add to the route/deploy contract for now, it's just personal preference, since in both case we will have to write the feed -> feed_id_type conversion function, either before the initialization of the new contract (@akhercha) or before adding to the associated router (@EvolveArt )

File: cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo (hunk 2 above)
Comment (akhercha) [reply cairo/crates/pragma_dispatcher/src/dispatcher/contract.cairo]: done

File: cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo (hunk 4 above)
Comment (akhercha) [reply cairo/crates/pragma_dispatcher/src/routers/feed_types/contracts/twap.cairo]: done

**Review by EvolveArt — APPROVED** [review]
lgtm
